### PR TITLE
Make some methods of `Pin` unstable const

### PR DESCRIPTION
**Pull request by CDirkx**
Make the following methods unstable const under the `const_pin` feature:
- `new`
- `new_unchecked`
- `into_inner`
- `into_inner_unchecked`
- `get_ref`
- `into_ref`
- `get_mut`
- `get_unchecked_mut`

Of these, `into_inner` and `into_inner_unchecked` require the unstable `const_precise_live_drops`.

Also adds tests for these methods in a const context.

Tracking issue: #76654

r? @ecstatic-morse

### CODE REVIEW

**Comment by CDirkx** [comment]
I remember checking if `Pin` methods could be made const before, however at that time I got an error that const fns could only be inside impl blocks that have `Sized` or no bounds on generic parameters. I couldn't find what enabled this?

**Comment by CDirkx** [comment]
I excluded `into_ref` (implemented on on `Pin<&mut T>`), which could technically also be made const, as `&mut` in consts is still being figured out I believe.

**Comment by ecstatic-morse** [comment]
> I remember checking if Pin methods could be made const before, however at that time I got an error that const fns could only be inside impl blocks that have Sized or no bounds on generic parameters. I couldn't find what enabled this?

This is currently part of the "min `const fn`" checks, which are disabled for const-unstable functions. Presumably you used `rustc_const_unstable` the last time you tried this.

> I excluded `into_ref` (implemented on on `Pin<&mut T>`), which could technically also be made const, as `&mut` in consts is still being figured out I believe.

I don't really see a reason to exclude this if it compiles. As long as these functions are all const-unstable, there's not really any risk.

cc @rust-lang/wg-const-eval Unlike `Option` and `Result`, I doubt `Pin` is useful in a const-context. However, making these unstably const might allow people to test things like const-qualification more thoroughly, and I don't see any downsides.

**Comment by CDirkx** [comment]
I added the methods `into_ref`, `get_mut`, `get_unchecked_mut` from `Pin<&mut T>`, the tests for these are a bit different.

File: library/core/tests/pin.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    const INNER_UNCHECKED: &'static usize = unsafe { Pin::into_inner_unchecked(PINNED) };
+    assert_eq!(INNER_UNCHECKED, POINTER);
+
+    const REF: &'static usize = PINNED.get_ref();
+    assert_eq!(REF, POINTER);
+
+    const fn pin_mut_const() {
+        // test that the methods of `Pin<&mut T>` are usable in a const context
+        // should be written as the other tests once `&mut` is usable in constants
```
Comment (RalfJung) [inline library/core/tests/pin.rs:24]: Are these two sentences? I think they are. If yes, please add capitalization and `.`. I had to read this 5 times to successfully parse it.

**Comment by ecstatic-morse** [comment]
r=me with Ralf's change.

@bors delegate+

**Comment by CDirkx** [comment]
r=@ecstatic-morse

**Comment by RalfJung** [comment]
Almost, the right command is

@bors r=ecstatic-morse

**Pull request merged**